text/3396-extern-types-v2.md

```diff
@@ -0,0 +1,249 @@
+}
+```
+This is the more obvious way of writing the trait, however it would prevent splitting `MetaSized` and `MetaAligned` in the future and adding these functions later would be hard but probably not impossible.
+
+
+# Prior art
+[prior-art]: #prior-art
+
+There is a lot of prior art in rust itself from previous attempts at custom DSTs, the `DynSized` trait, and some other things related to "exotically sized types".
+- [Lang team design notes on exotically sized types](https://github.com/rust-lang/lang-team/blob/master/src/design_notes/dynsized_constraints.md).  
+  This document contains notes from the lang team about what `?Sized + DynSized` needs to imply.
+  This document outlines `DynSized`, `MetaSized`, and `Sized` which inspired the "metadata sized" and friends in this RFC.
+  I believe this RFC satisfies the constraints outlined, mostly by dropping `DynSized` as an available bound, which limits expressiveness in favour of simplicity.
+- [Custom DSTs](https://github.com/rust-lang/rfcs/pull/2594).  
+  This postponed RFC attempts to introduce a generic framework for DSTs with arbitrary metadata.
+  This RFC aims to be compatible with future attempts at custom DSTs, as it is in essence a very restricted form.
+- [DynSized without ?DynSized](https://github.com/rust-lang/rfcs/pull/2310).  
+  This contains a lot of very useful analysis of what exactly commonly used crates want when they state `?Sized`.
+  However, this RFC aims to be simpler than the lint-based solution presented there.
+  Additionally, this deals with not being able to place extern types as fields in structs, rather than solely on `size_of_val` and `align_of_val`.
+- [More implicit bounds](https://github.com/rust-lang/rfcs/issues/2255).  
+  This discusses whether we should be adding more implicit bounds into the language, and there associated relaxations (like `?MetaSized`).
+  This RFC aims to sidestep this issue by utilising the edition system to change the definition of `?Sized`.
+
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+- Should "metadata sized" imply "metadata aligned" or should we be adding the `MetaAligned` trait rather than `MetaSized`?
+- Should `MetaSized` be a supertrait of `Sized`? All `Sized` things are `MetaSized` but `Sized` doesn't semantically require `MetaSized`.
+- Should users be able to slap a `#[repr(align(n))]` attribute onto opaque types to give them an alignment?  
+  This would allow us to represent `CStr` properly but would necessitate splitting `MetaSized` and `MetaAligned` as it is only "dynamically sized" but "statically aligned".
+  (We may be able to get away with the [Aligned trait](https://github.com/rust-lang/rfcs/pull/3319))
+- Should the `extern type` syntax exist, or should there just be a `repr(unsized)`?  
+  This would allow headers with opaque tails (which are very common in C code) but is a more significant departure from the original RFC, and looks more like custom DSTs.
```

Comment:
Why do you need `repr(unsized)` for this? Couldn't you just allow the last field of a type to be `?MetaSized`? (Perhaps requiring the struct to be `repr(C)`?)

Reply:
Imagine we have a setup like the following:
```rust
#[repr(C)]
struct Foo {
    a: u32,
    b: u8,
    foo: String,
}
#[repr(C)]
struct Bar {
    a: u32,
    b: u8,
    bar: u8,
}
extern type OpaqueTail;
#[repr(C)]
struct HeaderWithTail {
    a: u32,
    b: u8,
    tail: OpaqueTail,
}
#[repr(C, unsize)]
struct HeaderUnsized {
    a: u32,
    b: u8,
}
```
What alignment does `OpaqueTail` have? `&header.tail` doesn't really have a well defined meaning as it's either pointing at bar directly (assuming an alignment of one) or some random padding bytes before foo starts. However, you can freely convert between a `&Foo` (or `&Bar`) and a `&HeaderUnsize` without worrying about that.
This alignment issue is why this RFC doesn't propose allowing "dynamically aligned" or "unknown aligned" (`?MetaSized`) types as fields of structs (except as the only non-ZST field of a `#[repr(transparent)]` struct).

Reply:
Hmm, I suppose you could forbid referencing the field entirely…